### ISSUE TITLE
False positive on invalid-bytes-returned (E0308)

### ISSUE THREAD

**Issue by conqp** [PyCQA/pylint#3964]
### Steps to reproduce
Run pylint on the following code:

```
"""Common stuff facility."""

from __future__ import annotations
from hashlib import sha256
from json import dumps
from logging import getLogger
from os.path import getctime
from pathlib import Path
from sys import argv
from typing import NamedTuple, Union


__all__ = [
    'CHUNK_SIZE',
    'LOG_FORMAT',
    'LOGFILE',
    'LOGGER',
    'LOGGER',
    'sha256sum'
]


CHUNK_SIZE = 4 * 1024 * 1024    # Four Mebibytes.
LOG_FORMAT = '[%(levelname)s] %(name)s: %(message)s'
LOGFILE = Path('synclog.txt')
LOGGER = getLogger(Path(argv[0]).name)


class FileInfo(NamedTuple):
    """Stores meta information about a file."""

    sha256sum: str
    ctime: float

    def __bytes__(self) -> bytes:
        """Returns JSON-ish bytes."""
        return str(self).encode()

    def __str__(self) -> str:
        """Returns a JSON-ish string."""
        return dumps(self.to_json())

    @classmethod
    def from_file(cls, filename: Union[Path, str]) -> FileInfo:
        """Creates the file info from a file path."""
        return cls(sha256sum(filename), getctime(filename))

    def to_json(self) -> dict:
        """Returns JSON-ish dict."""
        return {'sha256sum': self.sha256sum, 'ctime': self.ctime}


def sha256sum(filename: Union[Path, str]) -> str:
    """Returns a SHA-256 sum of the specified file."""

    with open(filename, 'rb') as file:
        return sha256(file.read()).hexdigest()


def main():
    """Test FileInfo.__bytes__()."""

    file_info = FileInfo.from_file('/etc/fstab')
    print(bytes(file_info))


if __name__ == '__main__':
    main()
```

### Current behavior
```
$ pylint test.py
************* Module test
test.py:35:4: E0308: __bytes__ does not return bytes (invalid-bytes-returned)

------------------------------------------------------------------
Your code has been rated at 8.44/10 (previous run: 8.44/10, +0.00)

```

### Expected behavior
No warnings.

### pylint --version output
```
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.8.6 (default, Sep 30 2020, 04:00:38) 
[GCC 10.2.0]
```

**Comment by hippo91** [comment]
@conqp thanks for your report. It would be nice if you could give us the simplest snippet that is able to produce the bug.

**Comment by conqp** [comment]
Here's an even simpler one, that should™ not produce any pylint warnings:
```
#! /usr/bin/env python
"""Example of false-positive E0308."""

class String(str):
    """A string, convertible to UTF-8 bytes."""

    def __bytes__(self):
        return self.encode()
```

**Comment by hippo91** [comment]
@conqp thanks for your report. It is in fact the same issue as #3599.

**Comment by conqp** [comment]
It seems that way, yes.
But #3599 was reported in pylint 2.5.2 / astroid 2.4.1 and was fixed and closed months ago.
So why does it still occur in pylint 2.6.0 / astroid 2.4.2 now?

**Comment by hippo91** [comment]
@conqp the PR fixing #3599 has been merge on 06.23.20 whereas last astroid version (2.4.2) has been released on 06.08.20.
Thus the fix is not yet present in the distributed version but will be in the next one which should be released soon.
